src/pages/home/report/ReportActionCompose/SuggestionMention.tsx

```diff
@@ -300,16 +296,34 @@ function SuggestionMention(
 
     const getSuggestions = useCallback(() => suggestionValues.suggestedMentions, [suggestionValues]);
 
+    const onSelectionChange = useCallback(
+        (e: NativeSyntheticEvent<TextInputSelectionChangeEventData>) => {
+            /**
+             * we pass here e.nativeEvent.selection.end directly to calculateEmojiSuggestion
+             * because in other case calculateEmojiSuggestion will have an old calculation value
+             * of suggestion instead of current one
+             */
+            const selectionEnd = e.nativeEvent.selection.end;
+            // On iOS, onSelectionChange is called before onChangeText, so it's possible we calculate the mention suggestion with a new selection value but with an old composer value
+            // We use setTimeout to delay the calculation so both selection and value is matched correctly
+            setTimeout(() => {
+                calculateMentionSuggestionRef.current?.(selectionEnd);
+            }, 1);
+        },
+        [calculateMentionSuggestion],
+    );
```

Comment:
I will explain this change later.

Reply:
So, when I was testing after using the `selectionEnd`, I noticed one issue.

https://github.com/Expensify/App/assets/50919443/afcfc683-30ac-41ab-8c87-218184991a82

Test step performed on the above video
1. Type a mention
2. Add a character to the left of the mention (`a@bernhard`)
3. Remove the added character in 2 (`a|@bernhard`, press delete)

| = selection

What happens is the mention suggestion appears briefly. Why?

We currently have this `useEffect`, which will be triggered every time `selection` or `calculateMentionSuggestion` is updated.
https://github.com/Expensify/App/blob/0424a94e8c0c6d635429b812d6b0140b73ad8645/src/pages/home/report/ReportActionCompose/SuggestionMention.tsx#L281-L283

It's to fix this issue https://github.com/Expensify/App/issues/25649 where the selection value is updated first before the input value, so `calculateMentionSuggestion` is calculated with the new selection but with the old value, only on iOS. However, because the `useEffect` depends on `calculateMentionSuggestion` and `calculateMentionSuggestion` depends on `value`, the effect runs twice, once when the selection changes, and once when the `value` changes.

While it works "probably" fine on iOS, it affects the rest of the platform. On the rest of the platforms, the input `value` is updated first before the selection. So, based on the test step performed above, specifically the 3rd step, the value is updated to `@bernhard` but the selection is still 1, so the first `useEffect` shows the mention suggestion with the prefix of `@`, then the 2nd `useEffect` hides the mention suggestion.

So, I decided to move back again `calculateMentionSuggestion` to `onSelectionChange`, and to solve the "selection value is updated first before the input value" issue on iOS, I added a small timeout as the workaround. We use a `calculateMentionSuggestionRef` so the function is always up-to-date when called.

Reply:
This mismatch value and selection issue also happens on emoji suggestion, but it's unnoticable. Please watch the slowed down video below

https://github.com/Expensify/App/assets/50919443/b42ce0ee-5a6d-4f6d-a9a7-1e50d9bafd6b

1. Type `:smile`
2. Remove the `m`
3. The suggestion briefly shows the suggestion for `:si` even though the selection is in `s` (`:s|ile`)
| = selection

Let me know if we want to fix this for emoji suggestion too

Reply:
Thanks for investigating this issue.

This issue also exists in production. I went through long discussions in https://github.com/Expensify/App/issues/25649 which is a really tricky issue. I tend to believe it should be fixed from upstream. Given its complexity, I think this can also be considered as out of scope. Maybe we can log a separate issue to dig more about it instead of fixing it with `setTimeout` workaround here.

Reply:
I think we can focus on fixing the text selection highlight issue in this PR. 

Btw, I totally agree we should fix issues found during testing together if they have straightforward solutions. But for complex ones, I think it'll be better to handle them separately.